CORDA-941 contract verification

tudor-malene opened the pull request:
https://r3-cev.atlassian.net/browse/CORDA-941

This PR:

1. Adds a property: ``whitelistedContractImplementations`` to the ``NetworkParameters``
2. Creates a ``WhitelistedByZoneAttachmentConstraint`` that uses the above parameters to check that the attachment hashes in a transaction are found in the whitelist
3. Fixes the ``ContractAttachment`` to support multiple contracts
4. Update the ``DBAttachment`` to include the contract names, and the AttachmentService to return ``ContractAttachment``s
5. When creating a ``WireTransaction``, it will add the hashes of the current cordapps containing the contract used.

Basically, with this work, upgrading should be possible, but verification would always happen using the current contract implementation.

@mikehearn, there are 2 questions for you as comments. One in WireTransaction.kt and one in TransactionBuilder

Commit:
wire up and enforce max transaction size

tudor-malene committed:
CORDA-941 contract verification - WIP

tudor-malene committed:
CORDA-941 contract verification - revert ignored test

tudor-malene committed:
CORDA-941 contract verification - test fixes

tudor-malene committed:
CORDA-941 contract verification - ignore blocking test

tudor-malene committed:
CORDA-941 contract verification - added multiple contract names per ContractAttachment

tudor-malene committed:
CORDA-941 contract verification - fix tests

tudor-malene committed:
CORDA-941 contract verification - cleanup and fix tests

tudor-malene committed:
CORDA-941 contract verification - fix test

tudor-malene committed:
CORDA-941 contract verification - make ContractAttachment backward compatible

tudor-malene committed:
CORDA-941 contract verification - cleanup

tudor-malene committed:
CORDA-941 contract verification - api changes

tudor-malene committed:
CORDA-941 contract verification - fix test

tudor-malene committed:
CORDA-941 contract verification - fix tests

Commit:
Merge remote-tracking branch 'origin/master' into feature/CORDA-941/contract_verification_clean

tudor-malene committed:
CORDA-941 contract verification - more cleanup

tudor-malene committed:
CORDA-941 contract verification - refactoring and api fix

tudor-malene committed:
CORDA-941 contract verification - cleanup

tudor-malene committed:
CORDA-941 contract verification - fix

tudor-malene committed:
CORDA-941 contract verification - fix samples

Commit:
fixup after rebase
moved network parameter from AbstractNode to NodeProperties

Commit:
removed TODO

mikehearn commented on testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt:
```diff
@@ -64,10 +68,26 @@ class MockAttachmentStorage : AttachmentStorage, SingletonSerializeAsToken() {
+        val id = SecureHash.SHA256(hs.hash().asBytes())
+
+        if (!files.containsKey(id)) {
+            files[id] = Pair(ContractAttachment(object : AbstractAttachment({ bytes }) {
```
This would be a bit clearer if broken up into named sub-expressions.

tudor-malene commented on core/src/main/kotlin/net/corda/core/node/services/AttachmentStorage.kt:
```diff
@@ -52,6 +53,14 @@ interface AttachmentStorage {
+     * @param contractClassNames the list of contracts found in this attachment
+     * @return [AttachmentId] of uploaded attachment
+     */
+    fun importContractAttachment(contractClassNames: List<ContractClassName>, jar: InputStream): AttachmentId
```
I kept this mostly for performance reasons.
The CorDappLoader scanned the jar already and knows what contracts are in it, when it imports it. So, if using the normal import function, we'd have to rescan, which would add up for all the tests

mikehearn replied:
That's fair enough but I'd rather not complicate the API and open the possibility of new user mistakes, just to micro-optimise the unit test speed. We can add a backdoor later for testing purposes, that's internal API.

tudor-malene commented on core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt:
```diff
@@ -92,9 +92,7 @@ data class LedgerTransaction(
     private fun verifyConstraints() {
         val contractAttachments = attachments.filterIsInstance<ContractAttachment>()
         (inputs.map { it.state } + outputs).forEach { state ->
-            // Ordering of attachments matters - if two attachments contain the same named contract then the second
```
Aren't we enforcing a no-overlap rule on classloader building? 
So if 2 attachment have the same contract, it would fail when building the classloader

Commit:
fix broken import

tudor-malene commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -45,7 +46,8 @@ class NetworkBootstrapper {
         @JvmStatic
         fun main(args: Array<String>) {
-            val arg = args.singleOrNull() ?: throw IllegalArgumentException("Expecting single argument which is the nodes' parent directory")
+            val arg = args.singleOrNull()
```
I probably reformatted, and didn't realise that something else was changed.

Commit:
address PR issues

Commit:
remove API breaking change
address PR issue

tudor-malene commented on core/src/test/kotlin/net/corda/core/crypto/PartialMerkleTreeTest.kt:
```diff
@@ -58,7 +58,7 @@ class PartialMerkleTreeTest {
         expectedRoot = MerkleTree.getMerkleTree(hashed.toMutableList() + listOf(zeroHash, zeroHash)).hash
         merkleTree = MerkleTree.getMerkleTree(hashed)
-        testLedger = MockServices(emptyList(), rigorousMock<IdentityServiceInternal>().also {
+        testLedger = MockServices(listOf("net.corda.finance.contracts.asset"), rigorousMock<IdentityServiceInternal>().also {
```
Cool, I didn't know about that ``getCallerPackage`` method.

It's necessary because we use the contracts a bit earlier now. When the transaction is built.

tudor-malene commented on core/src/main/kotlin/net/corda/core/utilities/whitelistAllContractsForTest.kt:
```diff
@@ -0,0 +1,10 @@
+import net.corda.core.crypto.SecureHash
+
+/**
+ * This sequence can be used for test/demos
```
Oops, I forgot to clean up this bit.
Unfortunately it can't live in a test module because it has to be used in the core.

tudor-malene commented on core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt:
```diff
@@ -20,6 +21,11 @@ data class HashAttachmentConstraint(val attachmentId: SecureHash) : AttachmentCo
 }
 
+/** An [AttachmentConstraint] that verifies that the hash is in the network parameters whitelist*/
+data class WhitelistedByZoneAttachmentConstraint(val whitelistedContractImplementations: Set<SecureHash>) : AttachmentConstraint {
```
You're right, 
I rushed this, and didn't think it through properly.
Capturing the whitelist snapshot is clearly pointless.

Actually what we want this Constraint to do, is answer this question:

*"Is the classloader I'm using to verify the current transaction valid?
More precisely: For each transaction state, is the hash of the Jar file which provides the Contract implementation (and was used to form the classloader) whitelisted in the network parameters? "*

For 3.0, if we assume that all states are created with ``WhitelistedByZoneAttachmentConstraint``, and any contract and state changes are backward compatible, then we can keep using the existing CordappClassloader for the time being, and each time verify that the installed cordapps are whitelisted.

Commit:
Merge remote-tracking branch 'origin/pat/max-transaction-size' into feature/CORDA-941/contract_verification_clean

# Conflicts:
#	core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt
#	node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkMap.kt
#	node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt
#	testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt
#	testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt
#	testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/network/NetworkMapServer.kt
#	testing/test-common/src/main/kotlin/net/corda/testing/common/internal/ParametersUtilities.kt
#	tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt

tudor-malene committed:
CORDA-941 contract verification - code review changes

Commit:
Merge branch 'master' into feature/CORDA-941/contract_verification_clean

# Conflicts:
#	core/src/main/kotlin/net/corda/core/internal/GlobalProperties.kt
#	core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt
#	node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt
#	node/src/test/kotlin/net/corda/node/services/transactions/MaxTransactionSizeTests.kt
#	testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt
#	tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt

tudor-malene committed:
CORDA-941 contract verification - fix tests

tudor-malene committed:
CORDA-941 contract verification - fix api

tudor-malene commented:
Back to you @mikehearn .

I changed the approach so that it can be used with the current classloader.
1) I added a global flag: useWhitelistedByZoneAttachmentConstraint
2) On startup, if the flag is true, each deployed cordapp is checked against the whitelist ( and the node is failed if it’s invalid. Is this ok?)
3) if the flag is true, when building transactions, by default the ``WhitelistedByZoneAttachmentConstraint`` is used, instead of the ``HashAttachmentConstraint``
4) when creating the wire transaction, the attachment hashes of the current relevant cordapps are added to the Transaction.attachments ( as they are no longer in the HashAttachmentConstraint)
5) when verifying a transaction, the transaction attachment hashes are checked agains the current network whitelist 
6) based on the fact that we agreed that contracts are backwards compatible, the current installed cordapp should be able to verify any states coming from older transactions

Outstanding work: 
- add support to load the whitelist in the NetworkBootstrap 
- change the cordform gradle plugin to use the changed NetworkBootstrap
- document.

tudor-malene committed:
CORDA-941 contract verification - add contract whitelist support to the network bootstraper and cordform plugin

Commit:
Merge branch 'master' into feature/CORDA-941/contract_verification_clean

# Conflicts:
#	core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt
#	docs/source/example-code/src/test/kotlin/net/corda/docs/tutorial/testdsl/TutorialTestDSL.kt
#	node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt
#	testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/network/NetworkMapServer.kt

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt:
```diff
@@ -87,17 +87,20 @@ data class LedgerTransaction(
     /**
      * Verify that all contract constraints are valid for each state before running any contract code
      *
+     * In case the transaction was created on this node then the attachments will contain the hash of the current cordapp jars
```
Full stop at end of this sentence.

mikehearn commented on core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt:
```diff
@@ -83,31 +92,47 @@ open class TransactionBuilder(
+     * TODO - review this logic
+     */
+    private fun makeContractAttachments(services: ServicesForResolution): List<AttachmentId> {
+        fun getContractAttachmentId(state: TransactionState<ContractState>) = services.cordappProvider.getContractAttachmentID(state.contract)
```
This function could be inlined, right? Seems odd to name it.

tudor-malene replied:
Yes, I think initially it was used in multiple places.

mikehearn commented on docs/source/example-code/src/test/java/net/corda/docs/java/tutorial/testdsl/CommercialPaperTest.java:
```diff
@@ -32,7 +31,7 @@
     private static final TestIdentity MEGA_CORP = new TestIdentity(new CordaX500Name("MegaCorp", "London", "GB"));
     private final byte[] defaultRef = {123};
-    private final MockServices ledgerServices = new MockServices(MEGA_CORP);
+    private final MockServices ledgerServices = new MockServices(Arrays.asList("net.corda.finance.contracts"), MEGA_CORP);
```
Why is this change needed? Looks like maybe a merge conflict? The default c'tor is supposed to auto discover the package name.

tudor-malene replied:
It is needed because we need the cordapps now to add them to the WireTransaction.attachments.

Previously these tests, that were using the Transaction DSL, were wired with the ``AlwaysAcceptAttachmentConstraint``  constraint ( see ``TransactionDSL.input`` ), and thus the attachment hash was never needed.

AutoScanning cordapps only works if they are in the same package as the test and nothign is passed to the ``MockServices``. We usually pass in ``emptyList()``

mikehearn commented on docs/source/example-code/src/test/kotlin/net/corda/docs/tutorial/testdsl/TutorialTestDSL.kt:
```diff
@@ -38,7 +38,7 @@ class CommercialPaperTest {
     @JvmField
     val testSerialization = SerializationEnvironmentRule()
-    private val ledgerServices = MockServices(emptyList(), rigorousMock<IdentityServiceInternal>().also {
+    private val ledgerServices = MockServices(listOf("net.corda.finance.contracts"), rigorousMock<IdentityServiceInternal>().also {
```
Ditto  - seems unrelated to the PR

mikehearn commented on finance/src/test/java/net/corda/finance/contracts/asset/CashTestsJava.java:
```diff
@@ -38,7 +38,7 @@ public void trivial() {
         doReturn(MEGA_CORP.getParty()).when(identityService).partyFromKey(MEGA_CORP.getPublicKey());
         doReturn(MINI_CORP.getParty()).when(identityService).partyFromKey(MINI_CORP.getPublicKey());
-        transaction(new MockServices(emptyList(), identityService, MEGA_CORP.getName()), DUMMY_NOTARY, tx -> {
+        transaction(new MockServices(asList("net.corda.finance.contracts.asset"), identityService, MEGA_CORP.getName()), DUMMY_NOTARY, tx -> {
```
ditto: why have these all changed and is the auto-discovery sufficient?

mikehearn commented on gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Baseform.kt:
```diff
@@ -138,11 +138,13 @@ open class Baseform : DefaultTask() {
+        val bootstrapMethod = networkBootstrapperClass.getMethod("bootstrap", Path::class.java, List::class.java).apply { isAccessible = true }
         // Call NetworkBootstrapper.bootstrap
         try {
+            //create a list of all cordapps used in this network and pass it to the bootstrapper
```
Space after //, capital letters and full stop please.

mikehearn commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -47,12 +59,16 @@ class NetworkBootstrapper {
         fun main(args: Array<String>) {
-            val arg = args.singleOrNull() ?: throw IllegalArgumentException("Expecting single argument which is the nodes' parent directory")
-            NetworkBootstrapper().bootstrap(Paths.get(arg).toAbsolutePath().normalize())
+            val arg = args.firstOrNull()
```
Can you give `arg` a better name whilst you're here?

mikehearn commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -158,20 +176,48 @@ class NetworkBootstrapper {
+
+    private fun readContractWhitelist(file: Path): Map<String, List<AttachmentId>> = file.toFile().readLines()
+            .map { line -> line.split(":") }
+            .map { (contract, attachmenIds) ->
```
typo, attachmentIds is missing the t

mikehearn commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -158,20 +176,48 @@ class NetworkBootstrapper {
+    private fun readContractWhitelist(file: Path): Map<String, List<AttachmentId>> = file.toFile().readLines()
+            .map { line -> line.split(":") }
+            .map { (contract, attachmenIds) ->
+                contract to (attachmenIds.split(",").map { SecureHash.SHA256(it.parseAsHex()) })
```
SecureHash.parse is easier

mikehearn commented on node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt:
```diff
@@ -48,7 +48,7 @@ class BFTNotaryServiceTests {
     @Before
     fun before() {
-        mockNet = MockNetwork(emptyList())
+        mockNet = MockNetwork(listOf("net.corda.testing.contracts"))
```
likewise - you should be able to use a no-args c'tor here

mikehearn commented on node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt:
```diff
@@ -20,6 +25,32 @@ open class CordappProviderImpl(private val cordappLoader: CordappLoader, attachm
+            val whitelist = networkParameters.whitelistedContractImplementations
+                    ?: throw IllegalStateException("network parameters don't specify the whitelist")
+
+            //verify that the installed cordapps are whitelisted
```
comment style.

mikehearn commented on node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt:
```diff
@@ -20,6 +25,32 @@ open class CordappProviderImpl(private val cordappLoader: CordappLoader, attachm
+                (attch as ContractAttachment).allContracts.forEach { contractClassName ->
+                    val contractWhitelist = whitelist[contractClassName]
+                            ?: throw IllegalStateException("Contract ${contractClassName} is not whitelisted")
+                    if (attch.id !in contractWhitelist) IllegalStateException("Contract ${contractClassName} in attachment ${attch.id} is not whitelisted")
```
Can I have a better exception type than ISE please? Also I wonder if maybe a better solution is to log a warning or error rather than throw.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt:
```diff
@@ -85,7 +88,14 @@ class NodeAttachmentService(
+            @Column(name = "contract_class_name")
+            @CollectionTable(name = "node_attachments_contract_class_name", joinColumns = arrayOf(
+                    JoinColumn(name = "att_id", referencedColumnName = "att_id")),
+                    foreignKey = ForeignKey(name = "FK__ctr_class__attachments"))
```
Do fk names have to be very short for some reason? This name seems kind of ugly but I don't know db conventions well. Also how comes there are two underscores between some words but not others?

tudor-malene replied:
Yes, Oracle 11 has a 30 character identifier hard limit.

mikehearn replied:
What's up with the double underscores?

tudor-malene replied:
It's the convention we came up to for Foreign keys: FK__from_table__to_table
To differentiate from the underscores in the table names

mikehearn commented on testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt:
```diff
@@ -318,6 +321,8 @@ open class MockNetwork(private val cordappPackages: List<String>,
             return E2ETestKeyManagementService(identityService, keyPairs)
         }
 
+//        override fun retrieveNetworkParameters(trustRoot: X509Certificate) = super.retrieveNetworkParameters(trustRoot).copy(whitelistedContractImplementations = whitelistAllContractsForTest)
```
?

mikehearn reviewed (changes requested):
Yep this is looking much better.

shamsasari commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -47,12 +59,16 @@ class NetworkBootstrapper {
     }
 
-    fun bootstrap(directory: Path) {
+    fun bootstrap(directory: Path) = bootstrap(directory, null)
```
Is anything using this overload?

tudor-malene replied:
I kept it, in case it's used via reflection by some gradle plugin that I couldn't find.
But, now that I think about it, it's better it fails now during testing, so we spot the problem.

shamsasari commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -69,7 +85,7 @@ class NetworkBootstrapper {
             val notaryInfos = gatherNotaryInfos(nodeInfoFiles)
             println("Notary identities to be used in network-parameters file: ${notaryInfos.joinToString("; ") { it.prettyPrint() }}")
-            installNetworkParameters(notaryInfos, nodeDirs)
+            installNetworkParameters(notaryInfos, nodeDirs, if (cordapps != null) generateWhitelist(cordapps) else null)
```
`cordapps?.let { generateWhitelist(it) }` to be a bit more concise

shamsasari commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:
```diff
@@ -158,20 +176,48 @@ class NetworkBootstrapper {
     }
 
+    private fun generateWhitelist(cordapps: List<String>): Path {
+        val tempFile = Files.createTempFile("whitelist", ".txt")
```
Perhaps I'm missing something, but couldn't you directly create the mapping without writing to this temp file? `readContractWhitelist` parses what you've created here.

tudor-malene replied:
You're right, it's not necessary for our standard cordform use case.

I chose this approach as I was planning to reuse this generate function for the network operator tool to help setup the network parameters.
( Network admin would would take a list of cordapps and generate this whitelist, and then store it and use the tool again to append hashes and contracts to it. And use said whitelist when generating the parameters. Basically decoupling )

@shamsasari 
Does this make sense?

shamsasari replied:
As discussed offline the bootstrapper can simply read the old network-parameters file to get the existing whitelist.

shamsasari commented on node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt:
```diff
@@ -1,10 +1,15 @@
 import net.corda.core.cordapp.Cordapp
 import net.corda.core.cordapp.CordappContext
 import net.corda.core.crypto.SecureHash
+import net.corda.core.internal.GlobalProperties.networkParameters
```
ditto

shamsasari commented on testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt:
```diff
@@ -126,6 +126,10 @@ open class MockServices private constructor(
         }
+
+        init {
+            GlobalProperties.networkParameters = testNetworkParameters(emptyList())
```
What prompted this? A failing test? AbstractNode sets this on startup.

tudor-malene replied:
This line was added by @patrickkuo .
It's needed because networkParameters are global .

( In case you are wondering what @patrickkuo 's commits are doing here, it's because I merged from his branch before he pushed to master, to be able to refactor and use the GlobalProperties he introduced)

shamsasari replied:
So if you were to rebase onto master (or merge from master) this should go away?

mikehearn replied:
Can you rebase now Patrick's work is in master?

shamsasari commented on testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt:
```diff
@@ -56,18 +47,34 @@ class MockAttachmentStorage : AttachmentStorage, SingletonSerializeAsToken() {
-        }
-        catch (faee: java.nio.file.FileAlreadyExistsException) {
-            return AttachmentId.parse(faee.message!!)
+    private fun _importAttachment(jar: InputStream, contractClassNames: List<ContractClassName>? = null): AttachmentId {
```
`importAttachmentInternal`?

shamsasari commented on testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt:
```diff
@@ -24,30 +26,19 @@ class MockAttachmentStorage : AttachmentStorage, SingletonSerializeAsToken() {
-        return object : AbstractAttachment({ f }) {
-            override val id = id
-        }
+        return f.first
```
You can simplify this method further to just `files[id]?.first`

shamsasari commented on testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt:
```diff
@@ -56,18 +47,34 @@ class MockAttachmentStorage : AttachmentStorage, SingletonSerializeAsToken() {
+        val bytes = getBytes(jar)
+
+        val sha256 = bytes.sha256()
+        if (!files.containsKey(sha256)) {
```
You can use the `in` operator here

tudor-malene commented on core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt:
```diff
@@ -20,6 +23,25 @@ data class HashAttachmentConstraint(val attachmentId: SecureHash) : AttachmentCo
+     */
+    val whitelistAllContractsForTest = mapOf("*" to listOf(SecureHash.zeroHash, SecureHash.allOnesHash))
+
+    override fun isSatisfiedBy(attachment: Attachment) = networkParameters.whitelistedContractImplementations!!.let { whitelist ->
```
@shamsasari, 
it needs to be nullable because of serialisation backward compatibility.

tudor-malene commented on core/src/main/kotlin/net/corda/core/internal/GlobalProperties.kt:
```diff
@@ -4,6 +4,10 @@ import net.corda.core.node.NetworkParameters
 // TODO: This will cause problems when we run tests in parallel, make each node have its own properties.
 object GlobalProperties {
+
+    //todo - is this a reasonable approach?
```
I think that's a good idea.

tudor-malene commented on core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt:
```diff
@@ -20,6 +23,25 @@ data class HashAttachmentConstraint(val attachmentId: SecureHash) : AttachmentCo
+ * See: [net.corda.core.node.NetworkParameters.whitelistedContractImplementations]
+ * It allows for centralized control over the cordapps that can be used.
+ */
+object WhitelistedByZoneAttachmentConstraint : AttachmentConstraint {
```
Is it different from the other ``AttachmentConstraint(s)``?

tudor-malene committed:
CORDA-941 contract verification - address code review comments

Commit:
Merge branch 'master' into feature/CORDA-941/contract_verification_clean

# Conflicts:
#	core/src/test/kotlin/net/corda/core/transactions/LedgerTransactionQueryTests.kt
#	core/src/test/kotlin/net/corda/core/transactions/TransactionEncumbranceTests.kt
#	finance/src/test/java/net/corda/finance/contracts/asset/CashTestsJava.java
#	finance/src/test/kotlin/net/corda/finance/contracts/asset/ObligationTests.kt

tudor-malene committed:
CORDA-941 contract verification - fix unit tests

exFalso commented:
@mikehearn I haven't read the full PR, but from the relevant parts and the discussion we had the other day with @tudor-malene it seems to me that this way of doing the whitelisting poses a security threat.

In particular the whitelisting constraint's current definition doesn't include the *signing identity of attachment hashes*, which means Alice's signature over a transaction cannot be used to establish whether Alice trusts the signing identity. In other words this way of whitelisting makes the integrity of the ledger rely on trust that's not verifiable cryptographically based on the ledger alone.

The whitelisting-of-attachments problem mirrors the whitelisting-of-mrenclaves in sgx:
We must burn the whitelisting identity into the enclave in order to ensure that secrets are only revealed to whitelisted enclaves.
Similarly in this case we must burn the signing identity (the doorman or whatever pubkey) *into the transaction* (or rather the constraint) in order to ensure that only whitelisted contracts can spend the transaction's states.

If we don't do this then a signature over a transaction won't actually specify what whitelisting identity we trust.

So instead of the current design I'd suggest decoupling the constraint definition from the network map, the constraint should contain a pubkey, and the satisfiability check should simply check a signature of that identity over the attachment hash (or some whitelist datastructure). We would still need to figure out how to pass in the signature and the whitelist of course, perhaps we can slap it on the passed in Attachment?

tudor-malene committed:
CORDA-941 contract verification - fix unit tests

tudor-malene committed:
CORDA-941 contract verification - address code review comments

tudor-malene commented on core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt:
```diff
@@ -83,31 +92,47 @@ open class TransactionBuilder(
     fun toLedgerTransaction(services: ServiceHub) = toWireTransaction(services).toLedgerTransaction(services)
 
-    internal fun toLedgerTransactionWithContext(services: ServicesForResolution, serializationContext: SerializationContext) = toWireTransactionWithContext(services.cordappProvider, serializationContext).toLedgerTransaction(services)
+    internal fun toLedgerTransactionWithContext(services: ServicesForResolution, serializationContext: SerializationContext) =
```
It is long a one-liner.

shamsasari replied:
Sure it's long for one line. What I'm saying is use a code block (`{ ... }`) instead of just moving to the next line.

tudor-malene committed:
CORDA-941 contract verification - address code review comments

tudor-malene committed:
CORDA-941 contract verification - ignored test and added jira to fix

tudor-malene committed:
CORDA-941 contract verification - cleanup

tudor-malene commented:
@mikehearn , @shamsasari,
Back to you.

mikehearn commented:
@exFalso I think your point is a specific case of a more general question, which is should transactions specify the CZ to which they belong? In this case the identity of the signer over the list of whitelisted hashes is the identity of the CZ operator, so it's always 'clear' unless you have transactions that cross zones. 

It _does_ mean that a transaction using this feature does not fully encapsulate the state, and a transaction could cease being valid if the zone operator ever removes a whitelisting entry, which is one reason I don't really love this feature. But we have the same issue with notaries: if a notary is removed from the network parameters, then that also breaks the validity of old transactions. Solution: don't do that.

Now, should a tx include e.g. the hash of a network parameters document in it? It could. But in effect it already does via the notary pointers. As a notary cluster can belong to only a single zone at once, the notary a state is controlled by implicitly selects a set of network parameters too.

exFalso commented:
@mikehearn 
> But in effect it already does via the notary pointers. As a notary cluster can belong to only a single zone at once, the notary a state is controlled by implicitly selects a set of network parameters too.
But that's not a chain of trust we can check during verification!
I'll rephrase: how can we check this constraint inside SGX? The enclave would need to be supplied with the whitelist that's signed with a trusted identity, but we can't check whether that identity is trusted! A malicious host can easily sign a malicious attachment hash, there's nothing in the transaction that specifies the signing identity to check against!

mikehearn reviewed (changes requested):
Nearly there now.

mikehearn commented:
@exFalso Right but does it matter?

In SGX integrity mode, where we're checking the contents of a tx chain given to us by a remote peer, we're consuming the statement of validity ourselves. We have no incentive to supply the enclave with the wrong network parameters.

In SGX privacy mode, whereby we trust the remote attestation of a remote enclave to tell us that the transaction is valid, the attack you suggest is a real concern. But the enclave can include more than just the hash of the tx it checked in its statement of validity. It could sign a general purpose data structure that states "I checked tx of hash A using network parameters of hash B" using a key established during previous remote attestation which implicitly covered "... using enclave version C implicitly incorporating Corda platform version D with CPU of security version level E".

So the enclave ends up asserting multiple facts in its response, one of which can be the network parameters that were used to check the transaction. This fact _could_ also be in the tx, but only at the cost of bloating every single transaction with knowledge that is usually implicit in the execution environment.

kchalkias commented:
Had a discussion with @exFalso, his claims make some sense, but we might live with that.
a) One of the arguments that sounds interesting is: _Imagine you receive a copy of the ledger and you need to verify it's valid. As valid hashes are not reflected in the ledger, how do you know that at the time each tx was notarised the contract hash was indeed signed by the network map?_

The argument here is we trust the notary had the correct updated version at the time of check. As long as we trust the notary, we trust its clock, its updated whitelist etc.

BTW, there are a few other things we have to consider here, for example we don't check that NetworkMap params are signed by the NetworkMap entity itself (a Doorman's signature would also be accepted, do we want that?)

tudor-malene commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/ClassloaderUtils.kt:
```diff
@@ -0,0 +1,32 @@
+
+/**
+ * Scans the jar for contracts.
+ * @return: found contract class names or null if none found
```
Funnily, we use both.
@return ~ 300 times
@returns ~ 1000 times

shamsasari commented:
@kchalkias No, it is checked. See `net.corda.nodeapi.internal.network.NetworkMapKt#verifiedNetworkMapCert`

tudor-malene commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/ClassloaderUtils.kt:
```diff
@@ -0,0 +1,32 @@
+ */
+fun scanJarForContracts(cordappJarPath: String): List<ContractClassName>? {
+    val scanResult = FastClasspathScanner().addClassLoader(Thread.currentThread().contextClassLoader).overrideClasspath(cordappJarPath).scan()
+    val contracts = (scanResult.getNamesOfClassesImplementing(Contract::class.qualifiedName) + scanResult.getNamesOfClassesImplementing(UpgradedContract::class.qualifiedName)).distinct()
```
That's a good point. I took this code from the CordappLoader and assumed there was a reason for it.
I'll investigate

kchalkias commented:
@shamsasari great, so @exFalso we check against the cert role `CertRole.NETWORK_MAP`, no X500, no pubKey, no TLS assumptions. That said, this whitelist has passed the required signature verification tests before being stored. Due to this, the only hardcoded value is this NETWORK_MAP role and the ROOT_CERT in the truststore.
@mikehearn I am not sure if we've clarified if enclaves are CordaNetwork specific (thus root key is hardcoded) or they can work for any network, which might mean that the aforementioned keys/certs should be sent along with every tx as @exFalso mentioned.

tudor-malene committed:
CORDA-941 contract verification - address code review comments

Commit:
Merge branch 'master' into feature/CORDA-941/contract_verification_clean

# Conflicts:
#	gradle-plugins/cordformation/src/main/kotlin/net/corda/plugins/Node.kt
#	node-api/src/test/kotlin/net/corda/nodeapi/internal/AttachmentsClassLoaderStaticContractTests.kt
#	node-api/src/test/kotlin/net/corda/nodeapi/internal/AttachmentsClassLoaderTests.kt
#	node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt

exFalso commented:
@kchalkias @mikehearn To summarise the discussion and what we discussed offline:

1. Trust in the whitelist identity in the non-sgx use case is covered implicitly by the verifying notary signature, which should have the correct key in their truststore.
2. In the sgx case we have two designs, one catered for privacy of the backchain, one for integrity
 - Privacy use case: in this case the host may provide a malicious whitelist to the enclave, but there's no incentive to do so. Although come to think of it the host can use this to fish for information, i.e. pass different whitelists to reveal what attachments are used in the backchain.
 - Integrity use case: in this case we must establish the trust some other way inside the enclave. One way @mikehearn proposed is to require the certificate chain of the notary to be passed in, which should have the same root as the whitelist's signature. 
Another way that would cover both cases would be to burn the root key into the enclave itself, similarly to how we'll burn the enclave whitelisting key. This would piggyback on MRENCLAVE and tie each enclave image to the corresponding CZ. Which I think is fine.

kchalkias commented:
IMO the first "same-chain" suggestion cannot provide the required guaranties, eg. I can sign with a Node's key that share the same cert path with the notary. We should ensure it is indeed the NETWORK_MAP entity... atm, we do this:
```
fun <T : Any> SignedDataWithCert<T>.verifiedNetworkMapCert(rootCert: X509Certificate): T {
    require(CertRole.extract(sig.by) == CertRole.NETWORK_MAP) { "Incorrect cert role: ${CertRole.extract(sig.by)}" }
    X509Utilities.validateCertificateChain(rootCert, sig.by, rootCert)
    return verified()
}
```
Thus, first check the ROLE
Then check if I am one level away from the ROOT
I have my concerns on this as well (due to the inflexibility on adding intermediate levels, but we can tolerate that).
All in all, I am leaning towards burning the root key into the enclave.

mikehearn commented on core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt:
```diff
@@ -20,6 +23,27 @@ data class HashAttachmentConstraint(val attachmentId: SecureHash) : AttachmentCo
+ * See: [net.corda.core.node.NetworkParameters.whitelistedContractImplementations]
+ * It allows for centralized control over the cordapps that can be used.
+ */
+object WhitelistedByZoneAttachmentConstraint : AttachmentConstraint {
```
I'm going to side with Tudor here - it'd be weird if you could explore a public data structure (WireTransaction) and encounter internal-only objects in that graph. Are you even allowed to do `is` testing on such an object to ignore it? Perhaps not, as the object type could move or be renamed! This class is conceptually simple so I hope it isn't going to change after this PR goes in, so it should stay public.

mikehearn commented:
This PR was superseded by a different one against v3, and then another PR that brings it back to master. Closing.
